### PR TITLE
[Feat] Improve TPU CI

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR modify the TPU CI to rely on docker hub PyTorch Lightning xla-image, 
saving around 4 min, bringing CI time from 12-13 min to 8-9min

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: dockers/tpu-tests/tpu_test_cases.jsonnet [hunk 1]
```diff
@@ -10,17 +10,31 @@ local tputests = base.BaseTest {
 
   timeout: 900, # 15 minutes, in seconds.
 
-  image: std.extVar('image'),
-  imageTag: std.extVar('image-tag'),
+  image: 'pytorchlightning/pytorch_lightning:base-xla-py3.6-torch{PYTORCH_VERSION}',
+  imageTag: '{PR_NUMBER}-{PYTORCH_VERSION}',
```
Comment (zcain117) [inline dockers/tpu-tests/tpu_test_cases.jsonnet:14]: seems like `image` and `imageTag` are not used

File: .github/workflows/ci_test-tpu.yml [hunk 2]
```diff
@@ -16,81 +29,79 @@ env:
...
         gcloud container clusters get-credentials "$GKE_CLUSTER" --zone "$GKE_ZONE"
       shell: bash
 
     - name: Deploy the job on the kubernetes cluster
       env:
         XLA_VER: ${{ matrix.xla-version }}
       run: |-
-        python -c "fname = 'dockers/tpu-tests/tpu_test_cases.jsonnet' ; ttt = open(fname).read().replace('pytorch-VERSION', 'pytorch-$XLA_VER') ; open(fname, 'w').write(ttt)"
-        job_name=$(jsonnet -J ml-testing-accelerators/ dockers/tpu-tests/tpu_test_cases.jsonnet --ext-str image=$IMAGE --ext-str image-tag=$IMAGETAG | kubectl create -f -) && \
+        python -c "fname = 'dockers/tpu-tests/tpu_test_cases.jsonnet' ; ttt = open(fname).read().replace('{PYTORCH_VERSION}', '$XLA_VER').replace('{PR_NUMBER}', '${{ steps.vars.outputs.PR_NUMBER }}') ; open(fname, 'w').write(ttt)"
```
Comment (zcain117) [inline .github/workflows/ci_test-tpu.yml:102]: here I'd recommend passing PR number using `--ext-str pr_number=...` 

you might be able to use `--ext-str pr_number=$(steps.vars.outputs.PR_NUMBER)` or maybe you'd need to stick `steps.vars.outputs.PR_NUMBER` into an ENV var so you could use `--ext-str pr_number=PR_NUMBER`

File: dockers/tpu-tests/tpu_test_cases.jsonnet (hunk 1 above)
Comment (Borda) [reply dockers/tpu-tests/tpu_test_cases.jsonnet:14]: about commit SHA, won't be there collision if the user sends another build before the first ends?
also include python version...

File: dockers/tpu-tests/tpu_test_cases.jsonnet [hunk 3]
```diff
@@ -10,17 +10,31 @@ local tputests = base.BaseTest {
...
+      echo $PWD
+      git ls-remote --refs origin
+      git fetch origin "refs/pull/{PR_NUMBER}/head:pr/{PR_NUMBER}" && git checkout "pr/{PR_NUMBER}"
+
+      # drop horovod as it is not needed
+      python -c "fname = 'pytorch-lightning/requirements/extra.txt' ; lines = [line for line in open(fname).readlines() if not line.startswith('horovod')] ; open(fname, 'w').writelines(lines)"
+      # drop fairscale as it is not needed
+      python -c "fname = 'pytorch-lightning/requirements/extra.txt' ; lines = [line for line in open(fname).readlines() if 'fairscale' not in line] ; open(fname, 'w').writelines(lines)"
+
+      pip install -e .
```
Comment (Borda) [inline dockers/tpu-tests/tpu_test_cases.jsonnet:35]:
```suggestion
      pip install -e ".[dev]"
```
but why install itself?

File: dockers/tpu-tests/tpu_test_cases.jsonnet (hunk 1 above)
Comment (tchaton) [reply dockers/tpu-tests/tpu_test_cases.jsonnet]: I shouldn't need `imageTag` as my base image is directly pulled from DockerHub. https://hub.docker.com/r/pytorchlightning/pytorch_lightning/tags?page=1&ordering=last_updated&name=xla

File: dockers/tpu-tests/tpu_test_cases.jsonnet [hunk 4]
```diff
@@ -10,17 +10,31 @@ local tputests = base.BaseTest {
...
   accelerator: tpus.v3_8,
 
   command: utils.scriptCommand(
     |||
+      mkdir -p /home/runner/work/pytorch-lightning && cd /home/runner/work/pytorch-lightning
+      git clone https://github.com/PyTorchLightning/pytorch-lightning.git
       cd pytorch-lightning
+      echo $PWD
+      git ls-remote --refs origin
+      git fetch origin "refs/pull/{PR_NUMBER}/head:pr/{PR_NUMBER}" && git checkout "pr/{PR_NUMBER}"
```
[the remark this answers is not in the document]
Comment (tchaton) [reply dockers/tpu-tests/tpu_test_cases.jsonnet:27]: Any reasons why this would be needed, as the file is edited in the ci directly.

Screenshot 2021-02-22 at 09 55 36

File: dockers/tpu-tests/tpu_test_cases.jsonnet (hunk 1 above)
Comment (tchaton) [reply dockers/tpu-tests/tpu_test_cases.jsonnet]: Screenshot 2021-02-22 at 10 28 57

Comment (tchaton) [reply dockers/tpu-tests/tpu_test_cases.jsonnet]: Yes, I assumed it would work as one block. But jsonnet expect ImageTag separated from image name.

File: dockers/tpu-tests/tpu_test_cases.jsonnet (hunk 3 above)
Comment (tchaton) [reply dockers/tpu-tests/tpu_test_cases.jsonnet:31]: We need to re-install so we test the commit changes.

File: .github/workflows/ci_test-full.yml [hunk 5]
```diff
@@ -7,6 +7,7 @@ on:  # Trigger the workflow on push or pull request, but only for the master bra
   pull_request:
     branches: [master, "release/*"]
 
+
```
Comment (Borda) [inline .github/workflows/ci_test-full.yml:10]:
```suggestion
```

File: .github/workflows/ci_test-tpu.yml [hunk 6]
```diff
@@ -1,96 +1,104 @@
...
 jobs:
-  setup-build-publish-deploy:
+  setup-deploy:
     name: tpu-testing-job
     runs-on: ubuntu-20.04
     strategy:
       fail-fast: false
       matrix:
-        python-version: [3.6, 3.7]
+        python-version: [3.6]
```
Comment (Borda) [inline .github/workflows/ci_test-tpu.yml:34]: why not 3.7?

File: .github/workflows/ci_test-tpu.yml [hunk 7]
```diff
@@ -1,96 +1,104 @@
...
+        version: '290.0.1'
+        service_account_key: ${{ secrets.GKE_SA_KEY_BASE64 }}
+        project_id: ${{ secrets.GKE_PROJECT }}
+        export_default_credentials: true
+
+    - name: Checkout Pytorch Lightning
+      uses: actions/checkout@v2
+      with:
+        repository: PyTorchLightning/pytorch-lightning
+        ref: ${{ github.event.base_ref }}
```
Comment (Borda) [inline .github/workflows/ci_test-tpu.yml:52]: why, by default it checkout the current PR/commit

File: .github/workflows/ci_test-tpu.yml [hunk 8]
```diff
@@ -1,96 +1,104 @@
...
       shell: bash
 
     - name: Deploy the job on the kubernetes cluster
       env:
         XLA_VER: ${{ matrix.xla-version }}
+        PYTHON_VER: ${{ matrix.xla-version }}
       run: |-
-        python -c "fname = 'dockers/tpu-tests/tpu_test_cases.jsonnet' ; ttt = open(fname).read().replace('pytorch-VERSION', 'pytorch-$XLA_VER') ; open(fname, 'w').write(ttt)"
-        job_name=$(jsonnet -J ml-testing-accelerators/ dockers/tpu-tests/tpu_test_cases.jsonnet --ext-str image=$IMAGE --ext-str image-tag=$IMAGETAG | kubectl create -f -) && \
+        python -c "fname = 'dockers/tpu-tests/tpu_test_cases.jsonnet' ; ttt = open(fname).read().replace('{PYTORCH_VERSION}', '$XLA_VER').replace('{PYTHON_VERSION}', '$PYTHON_VER').replace('{PR_NUMBER}', '${{ steps.vars.outputs.PR_NUMBER }}') ; open(fname, 'w').write(ttt)"
```
Comment (Borda) [inline .github/workflows/ci_test-tpu.yml:99]: lets move it to python step so it readable

File: dockers/tpu-tests/tpu_test_cases.jsonnet [hunk 9]
```diff
@@ -10,17 +10,27 @@ local tputests = base.BaseTest {
...
   command: utils.scriptCommand(
     |||
+      source ~/.bashrc
+      conda activate lightning
+      mkdir -p /home/runner/work/pytorch-lightning && cd /home/runner/work/pytorch-lightning
+      git clone https://github.com/PyTorchLightning/pytorch-lightning.git
       cd pytorch-lightning
+      echo $PWD
+      git ls-remote --refs origin
+      git fetch origin "refs/pull/{PR_NUMBER}/head:pr/{PR_NUMBER}" && git checkout "pr/{PR_NUMBER}"
```
Comment (Borda) [inline dockers/tpu-tests/tpu_test_cases.jsonnet:30]: can we get here rather commit hash, at least to verify, since in extreme case in the delay between creating this docker and trigger action there can be another commit...

File: .github/workflows/ci_test-full.yml (hunk 5 above)
Comment (tchaton) [reply .github/workflows/ci_test-full.yml:10]: Done !

File: .github/workflows/ci_test-tpu.yml (hunk 6 above)
Comment (tchaton) [reply .github/workflows/ci_test-tpu.yml:34]: I am putting it back, but running a matrix of 4 seems quite a lot to me.

File: .github/workflows/ci_test-tpu.yml (hunk 7 above)
Comment (tchaton) [reply .github/workflows/ci_test-tpu.yml:52]: I didn't know. Thanks !

File: .github/workflows/ci_test-tpu.yml (hunk 8 above)
Comment (tchaton) [reply .github/workflows/ci_test-tpu.yml]: Done !

File: dockers/tpu-tests/tpu_test_cases.jsonnet (hunk 9 above)
Comment (tchaton) [reply dockers/tpu-tests/tpu_test_cases.jsonnet:30]: Done !

File: .circleci/config.yml [hunk 10]
```diff
@@ -33,25 +46,28 @@ references:
...
+      name: Update jsonnet
       command: |
-        gcloud --quiet auth configure-docker
-        #cd dockers/tpu-tests
-        docker build --tag "$GCR_IMAGE_PATH:$CIRCLE_WORKFLOW_JOB_ID" -f ./dockers/tpu-tests/Dockerfile --build-arg "PYTHON_VERSION=$PYTHON_VER" --build-arg "PYTORCH_VERSION=$XLA_VER" .
-        docker push "$GCR_IMAGE_PATH:$CIRCLE_WORKFLOW_JOB_ID"
+       export PR_NUMBER=$(git ls-remote origin "pull/*/head" | grep -F -f <(git rev-parse HEAD) | awk -F'/' '{print $3}')
+       export SHA=$(git rev-parse --short HEAD)
+       python -c "fname = 'dockers/tpu-tests/tpu_test_cases.jsonnet' ; fff = open(fname).read().replace('{PYTORCH_VERSION}', '$XLA_VER')
+       fff = fff.replace('{PYTHON_VERSION}', '$PYTHON_VER').replace('{PR_NUMBER}', '$PR_NUMBER').replace('{SHA}', '$SHA') ; open(fname, 'w').write(fff)"
```
Comment (tchaton) [inline .circleci/config.yml:62]:
```suggestion
       fff.replace('{PYTHON_VERSION}', '$PYTHON_VER').replace('{PR_NUMBER}', '$PR_NUMBER').replace('{SHA}', '$SHA') ; open(fname, 'w').write(fff)"
```

**Review by tchaton — COMMENTED** [review]
LGTM!

**Review by awaelchli — APPROVED** [review]

File: .circleci/config.yml (hunk 10 above)
Comment (carmocca) [reply .circleci/config.yml:62]: `str.replace` is not inplace

File: .circleci/config.yml [hunk 11]
```diff
@@ -33,25 +46,28 @@ references:
...
-        - PYTHON_VER: 3.7
+      name: Update jsonnet
       command: |
-        gcloud --quiet auth configure-docker
-        #cd dockers/tpu-tests
-        docker build --tag "$GCR_IMAGE_PATH:$CIRCLE_WORKFLOW_JOB_ID" -f ./dockers/tpu-tests/Dockerfile --build-arg "PYTHON_VERSION=$PYTHON_VER" --build-arg "PYTORCH_VERSION=$XLA_VER" .
-        docker push "$GCR_IMAGE_PATH:$CIRCLE_WORKFLOW_JOB_ID"
+       export PR_NUMBER=$(git ls-remote origin "pull/*/head" | grep -F -f <(git rev-parse HEAD) | awk -F'/' '{print $3}')
+       export SHA=$(git rev-parse --short HEAD)
+       python -c "fname = 'dockers/tpu-tests/tpu_test_cases.jsonnet' ; fff = open(fname).read().replace('{PYTORCH_VERSION}', '$XLA_VER')
```
Comment (carmocca) [inline .circleci/config.yml:61]: what's `fff` supposed to mean?

Comment (kaushikb11) [reply .circleci/config.yml:61]: @carmocca It's just the variable/placeholder for jsonnet data.

Comment (carmocca) [reply .circleci/config.yml:61]: but why `fff` and not `cases` or `data`

Comment (kaushikb11) [reply .circleci/config.yml]: Updated!

**Review by carmocca — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]
